enhancements/coreos-encrypted-by-default.md

```diff
@@ -0,0 +1,422 @@
+
+### Non-Goals
+
+The following are considered out of scope:
+* Encryption of installations not based on a compatible boot image.
+* Changing Ignition.
+* Installer support.
+* Storing of keys in-Cluster.
+* Arbitrary LUKS configurations or disk-layouts.
+* Re-formatting disk devices or partitions.
+
+### Fedora (FCOS) and Red Hat CoreOS (RHCOS)
+
+It is worth stating that this proposal is not mutually exclusive for any plans that FCOS may have towards root disk layout. FCOS and RHCOS would like to reach alignment and feature parity.
+
+With the exclusion of the on-disk format, this work should lay the groundwork for an unattended encrypted boot that can be used by both FCOS and RHCOS.
+
+Ignition is the first boot agent for FCOS and RHCOS. FCOS discussed the topic of partition layouts, with the idea of having Ignition lay-down arbitrary root filesystem layouts. At this time, Openshift requires Ignition V2, while FCOS uses Ignition v3; these versions are incompatible. When Ignition supports the required depenendencies, a reviewion to this proposal will be forthcoming.
+
+## Terms
+* LUKS/LUKS2: the standard Linux encryption method. Encrypted filesystems stored in a container.
+* Master Key: the _actual_ encryption key used. The master key encrypted using a passphrase and is embedded in the LUKS meta-data on disk encrypted. This encryption key is unknown to users.
+* LUKS. Passphrase: any string or file that can be used to unlock the master key. Each passphrase is bound to a "key-slot."
+* LUKS Token: arbitrary JSON stored in the LUKS meta-data.
+* NBE: [Network Bound-Disk Encryption](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/Security_Guide/sec-Using_Network-Bound_Disk_Encryption.html). A method of providing context-based disk encryption.
+* [Clevis](https://github.com/latchset/clevis): Linux CLI tools for policy-based encryption and decryption. Clevis has a Dracut module for automatically opening LUKS containers in an NBE context.
+* Clevis pin: a program called by Clevis to encrypt and decryption of secrets.
+* Tang: A server implementation that provides cryptographic binding services without the need for an escrow. Secrets are derived from what the client and the Tang server both have, reducing the attack surface.
+* KMS: Key Management System, such as Amazon KMS.
+* [TPM2](https://en.wikipedia.org/wiki/Trusted_Platform_Module): A secure cryptoprocessor for storing keys in hardware.
+* Dracut Module: program(s) that run early in boot before the on-disk Linux is booted.
+
+If you are unfamiliar with some of these terms [please see an excellent Youtube video on Clevis/Tang](https://youtu.be/2uLKvB8Z5D0).
+
+## Proposal
+
+To provide install-time _and_ rolling re-encryption, CoreOS boot images will:
+* Have the root filesystem housed in a LUKS container.
+* Use the 'null-cipher' for the initial LUKS container "encrpytion". This allows _any_ passphrase to unlocked the LUKS container.
```

Comment:
This means people who don't want/need encryption are paying at least some cost.  Worth measuring to see what that is.  Also, part of the cost is things like [disabling TRIM by default](https://wiki.archlinux.org/index.php/Dm-crypt/Specialties#Discard/TRIM_support_for_solid_state_drives_(SSD)).  (And while people still debate how and when to `TRIM`...doing so can be a *major* performance improvement)

Reply:
I've benchmarked that -- the cost is negligible. I was unable to find a statistically significant cost in the null-cipher case.

Reply:
I think that the answer this critique is simple: we add the kargs of `rd.luks.options=discard` or setting the persistent of during the initial build time would work. In the cloud use case (where users pay by storage allocated) enabling discard makes a lot of sense. I lean toward using the karg method. 

That should negate the user costs in terms of money. The performance cost, as I noted elsewhere is negligible.